Requirements.md

```diff
@@ -0,0 +1,101 @@
+   
+   For example, if a type cast is potentially eliminatable, then the engine should not be required to discover that fact; rather the toolchain should be able to eliminate the type cast.
+   
+1. Support code migration.
+
+   This refers to a potential code migration strategy; where an application starts as a JavaScript code base (say), and is then progressively migrated to another language (hosted on WASM-GC). During this migration process, individual functions and modules may be ported from JavaScript whilst the remaining codebase remains in JavaScript. 
+   Interoperability between WASM-GC code and JavaScript is critical for the success of this strategy; in particular, structures defined in one language may need to be accessed by another.
+   Code Migration itself is a CSF because it underlies a realistic assumption about porting libraries and applications to WASM-GC.
+
+## Requirements
+
+1. Support for typical memory layouts employed by languages like Java/C#/Kotlin.
+
+1. Support for V-tables for object layout.
+
+1. Support for embedded arrays.
+
+   Many languages use a combined structure consisting of a fixed set of fields together with an array. This is, for example, how arrays themselves are realized in languages like Java.
+   
+   Potentially restricted to a single embedded array.
+   
+1. Support for parametric polymorphic types.
+   Specifically, support for languages that support parametric type polymorphism; and subtype polymorphism. 
+   The primary task of the WASM type system is to be able to reliably and securely identify the representations of values. As such, it is not clear that WASM's type system needs to be polymorphic. However, many statically typed languages are polymorphic, and there are multiple strategies for realizing them -- from a so-called uniform representation (for representing parametric types) to compile-time monomorphization.
+   
+
+### Non-Requirements
+1. Arbitrary interoperability between different languages on the wasm.
+   While it may be tempting to try to support arbitrary interoperability across languages -- by enabling cross language access to data structures in the case of WASM-GC -- this is fraught with many difficulties; not the least of which are caused by those languages themselves.
+   The task of WASM-GC is to enable managed languages to represent their structures in such a way as can be managed by the host. This does not imply interoperability between languages or between languages and the host.
+1. Arbitrary interoperability with host.
+   Similar to the above issue.
+1. Permit access to host data structures from wasm, and vice versa.
+   While it may be desireable to support access to host structures from wasm, and access to wasm structures from the host, it is not the goal of WASM-GC to support this.
```

Comment:
@dcodeIO: 

> But at least WebAssembly and JavaScript must work well together, right, so why not just write down the obvious?

This is not at all obvious IMO. I phrased it in the document as supporting code migration; because that is a clear use case.

The fundamental issue is that there is a semantic mismatch (only one?) between JS and WASM: JS is a user facing programming language and WASM is only ever the target of compilers.

What that means is that we cannot reasonably talk about WASM<->JS interop; we must talk about XX<->JS interop, where XX is one of many languages that compile to WASM. 

Simple example, JS supports named fields of objects that can have dynamically installed getters and setters. Not all languages support named fields, and not all languages (pretty few actually) support dynamically installed getters and setters.

Reply:
> What that means is that we cannot reasonably talk about WASM<->JS interop; we must talk about XX<->JS interop, where XX is one of many languages that compile to WASM.

But good XX<->(Wasm_compiler_target<->)JS_user_facing_programming_language interop will only be a viable endeavour if there is good Wasm<->JS interop to start with, no matter if XX is already reasonably JS compatible or decides to build additional abstractions because they see advantages in doing so. What's "not at all obvious" about that?